tests/features/S2N_DIAGNOSTICS_POP_SUPPORTED.c

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/**
+ * This feature detects if the compiler properly pops diagnostics
+ */
+
+#include <stdint.h>
+
+#define MACRO_CHECK \
+    do { \
+        _Pragma("GCC diagnostic push") \
+        _Pragma("GCC diagnostic error \"-Wconversion\"") \
+        return -1; \
+        _Pragma("GCC diagnostic pop") \
+    } while (0)
+
+int signed_fun()
+{
+    MACRO_CHECK;
+}
+
+/* this function is here to ensure the compiler properly pops the previous diagnostic */
+uint8_t unsigned_fun()
+{
+    return -1;
+}
```

Comment:
fun fact: <GCC5 doesn't implement `diagnostic pop` correctly and preserves the configuration for the rest of the file: https://godbolt.org/z/qdKrhWMo6

Reply:
Is there ever an instance where we would want to use GCC diagnostics where `diagnostic pop` isn't supported? Like, should checking if `diagnostic pop` works correctly just be a part of the normal GCC diagnostic feature probe?

Reply:
Great question (and this ties into your other question). There's two different uses of the diagnostics in the codebase: you can ignore warnings that are on by default and you can turn on warnings that are off by default. In the case of ignoring warnings, we want to include as many compilers in that list as possible, otherwise we'd get warnings on those platforms and fail the build (since `-Werror` is on). The worst case is we ignore warnings later on, _but_ since we're also testing on platforms that do support popping diagnostics, it's not going to affect those.

On the other hand, we definitely can't turn on warnings for compilers that don't support pop without fixing every instance in the file after that triggers the warning.